File: js/src/ast.ts
export interface Document {
  type: 'Document';
  entries: GomlNode[];
}

export interface KeyValue {
  type: 'KeyValue';
  key: string;
  value: GomlValue;
  line: number;
  col: number;
}

export interface BlockObject {
  type: 'Object';
  name: string;
  entries: GomlNode[];
  line: number;
  col: number;
}

export interface BlockArray {
  type: 'Array';
  name: string;
  items: GomlValue[];
  line: number;
  col: number;
}

export interface CommentNode {
  type: 'Comment';
  text: string;
  line: number;
  col: number;
}

export interface StringValue {
  type: 'String';
  value: string;
  line: number;
  col: number;
}

export interface NumberValue {
  type: 'Number';
  value: number;
  raw: string;
  line: number;
  col: number;
}

export interface BooleanValue {
  type: 'Boolean';
  value: boolean;
  line: number;
  col: number;
}

export interface NullValue {
  type: 'Null';
  line: number;
  col: number;
}

export interface ArrayValue {
  type: 'ArrayValue';
  items: GomlValue[];
  line: number;
  col: number;
}

export interface InlineObject {
  type: 'InlineObject';
  entries: KeyValue[];
  line: number;
  col: number;
}

export interface ReferenceValue {
  type: 'Reference';
  path: string;
  line: number;
  col: number;
}

export type GomlValue =
  | StringValue
  | NumberValue
  | BooleanValue
  | NullValue
  | ArrayValue
  | InlineObject
  | ReferenceValue;

export type GomlNode = KeyValue | BlockObject | BlockArray | CommentNode;

File: internal/ast/ast.go
package ast

type Node interface {
	nodeType() string
}

type Document struct {
	Statements []Node
	Line       int
	Col        int
}

func (d *Document) nodeType() string { return "Document" }

type Comment struct {
	Text string
	Line int
	Col  int
}

func (c *Comment) nodeType() string { return "Comment" }

type KeyValue struct {
	Key   string
	Value Node
	Line  int
	Col   int
}

func (kv *KeyValue) nodeType() string { return "KeyValue" }

type Object struct {
	Entries []Node
	Line    int
	Col     int
}

func (o *Object) nodeType() string { return "Object" }

type Array struct {
	Items []Node
	Line  int
	Col   int
}

func (a *Array) nodeType() string { return "Array" }

type StringVal struct {
	Value string
	Line  int
	Col   int
}

func (s *StringVal) nodeType() string { return "String" }

type NumberVal struct {
	Value    string
	IsFloat  bool
	FloatVal float64
	IntVal   int
	Line     int
	Col      int
}

func (n *NumberVal) nodeType() string { return "Number" }

type BoolVal struct {
	Value bool
	Line  int
	Col   int
}

func (b *BoolVal) nodeType() string { return "Boolean" }

type NullVal struct {
	Line int
	Col  int
}

func (n *NullVal) nodeType() string { return "Null" }

type Reference struct {
	Path string
	Line int
	Col  int
}

func (r *Reference) nodeType() string { return "Reference" }
